chirp/drivers/tk690.py

```diff
@@ -0,0 +1,1340 @@
+            "1 - Turn off your radio\n"
+            "2 - Connect your interface cable\n"
+            "3 - Turn on your radio (unlock it if password protected)\n"
+            "4 - Click OK to start\n")
+        return rp
+
+    def get_features(self):
+        """Return information about this radio's features"""
+        rf = chirp_common.RadioFeatures()
+        rf.has_settings = True
+        rf.has_bank = True
+        rf.has_tuning_step = False
+        rf.has_name = True
+        rf.has_offset = True
+        rf.has_mode = True
+        rf.has_dtcs = True
+        rf.has_rx_dtcs = True
+        rf.has_dtcs_polarity = True
+        rf.has_ctone = True
+        rf.has_cross = True
+        rf.valid_modes = MODES
+        rf.valid_duplexes = ["", "-", "+", "off"]
+        rf.valid_tmodes = ['', 'Tone', 'TSQL', 'DTCS', 'Cross']
+        rf.valid_cross_modes = [
+            "Tone->Tone",
+            "DTCS->",
+            "->DTCS",
+            "Tone->DTCS",
+            "DTCS->Tone",
+            "->Tone",
+            "DTCS->DTCS"]
+        rf.valid_power_levels = POWER_LEVELS
+        rf.valid_characters = VALID_CHARS
+        rf.valid_skips = SKIP_VALUES
+        rf.valid_dtcs_codes = DTCS_CODES
+        rf.valid_bands = [self._range]
+        rf.valid_name_length = NAME_CHARS
+        rf.memory_bounds = (1, self._upper)
+        return rf
+
+    def _fill(self, offset, data):
+        """Fill an specified area of the memmap with the passed data"""
+        for addr in range(0, len(data)):
+            self._mmap[offset + addr] = data[addr]
+
+    def _prep_data(self):
+        """Prepare the areas in the memmap to do a consistend write
+        it has to make an update on the x1600 area with banks and channel
+        info; other in the x1000 with banks and channel counts
+        and a last one in x7000 with flog data"""
+        rchs = 0
+        data = dict()
+
+        # sorting the data
+        for ch in range(0, self._upper):
+            mem = self._memobj.memory[ch]
+            bnumb = int(mem.bnumb)
+            bank = int(mem.bank)
+            if bnumb != 255 and (bank != 255 and bank != 0):
+                try:
+                    data[bank].append(ch)
+                except KeyError:
+                    data[bank] = list()
+                    data[bank].append(ch)
+                data[bank].sort()
+                # counting the real channels
+                rchs = rchs + 1
+
+        # updating the channel/bank count
+        self._memobj.settings.channels = rchs
+        self._chs_progs = rchs
+        self._memobj.settings.banks = len(data)
+
+        # building the data for the memmap
+        fdata = b""
+
+        for k, v in data.iteritems():
+            # posible bad data
+            if k == 0:
+                k = 1
+                raise errors.InvalidValueError(
+                    "Invalid bank value '%d', bad data in the image? \
+                    Triying to fix this, review your bank data!" % k)
```

Comment:
I'm looking through this _prep_data function, as far as I can find it's not actually invoked by the code at all? It also had some clearly nonsensical aspects to it earlier on, trying to write some values to 0x7000 while the memory in this radio only goes up until 0x60A0. Unless you see any valuable functionality in it that I'm overlooking, making any of this logic worth keeping around somewhere, I'm thinking I'll just get rid of the whole function (then there's no need to worry about any confusing logic within it lol)

Reply:
Yeah, I don't see it called either. Frustrating :(

One pattern you'll see often is people copying a driver and modifying it to work for a different radio. That ends up with all kinds of orphaned crap, and this could be a similar situation. With hundreds of drivers to look at, it's quite hard to notice and reject those things so plenty slipped in over the years. There's a job in "PR checks" that tries to spot overly-similar drivers now, but it's not perfect of course.

So yes, please nuke it if it's not useful/used.